src/dynamic-launcher.c

```diff
@@ -0,0 +1,868 @@
+
+  g_hash_table_insert (transient_permissions,
+                       g_strdup (token),
+                       g_variant_ref_sink (launcher_data_wrapped));
+}
+
+static void
+prepare_install_done (GObject      *source,
+                      GAsyncResult *result,
+                      gpointer      data)
+{
+  g_autoptr(Request) request = data;
+  GVariant *launcher_data;
+  guint response = 2;
+  xdp_autofd int icon_fd = -1;
+  g_autoptr(GVariant) results = NULL;
+  g_autoptr(GError) error = NULL;
+  GVariantBuilder results_builder;
+
+  REQUEST_AUTOLOCK (request);
+
+  g_variant_builder_init (&results_builder, G_VARIANT_TYPE_VARDICT);
+
+  if (!xdp_impl_dynamic_launcher_call_prepare_install_finish (XDP_IMPL_DYNAMIC_LAUNCHER (source),
+                                                              &response,
+                                                              &results,
+                                                              result,
+                                                              &error))
+    {
+      g_warning ("Backend call failed: %s", error->message);
+      goto out;
+    }
+
+  if (request->exported && response == 0)
+    {
+      g_autofree char *token = g_uuid_string_random ();
+      const char *chosen_name = NULL;
+      const char *icon_format = NULL;
+      GVariant *chosen_icon = NULL;
+
+      icon_format = g_object_get_data (G_OBJECT (request), "icon-format");
+      g_assert (icon_format);
+
+      if (!xdp_filter_options (results, &results_builder,
+                               response_options, G_N_ELEMENTS (response_options),
+                               &error) ||
+          !g_variant_lookup (results, "name", "&s", &chosen_name) ||
+          !g_variant_lookup (results, "icon", "v", &chosen_icon))
+        {
+          g_warning ("Results from backend failed validation: %s",
+                     error ? error->message : "missing entries");
+          response = 2;
+        }
+      else
+        {
+          /* Save the token in memory and return it to the caller */
+          launcher_data = g_variant_new ("(svs)", chosen_name, chosen_icon, icon_format);
+          set_launcher_data_for_token (token, launcher_data);
+          g_variant_builder_add (&results_builder, "{sv}", "token", g_variant_new_string (token));
+        }
+    }
+
+out:
+  if (request->exported)
+    {
+      xdp_request_emit_response (XDP_REQUEST (request),
+                                 response,
+                                 g_variant_builder_end (&results_builder));
+
+      request_unexport (request);
+    }
+  else
+    {
+      g_variant_builder_clear (&results_builder);
+    }
+}
+
+static gboolean
+validate_url (const char  *key,
+              GVariant    *value,
+              GVariant    *options,
+              GError     **error)
+{
+  const char *url = g_variant_get_string (value, NULL);
+  g_autoptr(GError) local_error = NULL;
+
+  if (!g_uri_is_valid (url, G_URI_FLAGS_NONE, &local_error))
```

Comment:
Is this correct for the non-webapp case?

Reply:
You're right, we need to only do this validation in the webapp case.